nbformat/v4/nbformat.v4.schema.json

```diff
@@ -76,6 +76,40 @@
                       },
                       "additionalProperties": true
                     }
+                },
+                "environments": {
+                    "description": "Collection of environments in which this notebook was authored and executed",
+                    "type": "array",
+                    "item": {
+                        "description": "Specification of an environment in which this notebook was authored and executed",
+                        "type": "object",
+                        "anyOf": [
+                            {"required": ["type", "source"]},
+                            {"required": ["type", "spec"]}
+                        ],
+                        "properties": {
+                            "type": {
+                                "description": "The type of environment specification, preferably a mimetype including a version number (e.g., application/vnd.docker.dockerfile.v1)",
+                                "type": "string"
+                            },
+                            "source": {
+                                "description": "URI of an environment specification in the format defined by 'type', stored external to the notebook document. Mutually exclusive with spec.",
+                                "type": "string"
+                            },
+                            "spec": {
+                                "description": "Environment specification in the format defined by 'type', stored within the notebook document itself. Mutually exclusive with source."
```

Comment:
Should 'spec' have a requirement of 'type: object'? Or are we allowing it to be any on purpose for things like a pip requirements list being an Array rather than an object?